resources/js/components/textarea-autosize.tsx

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import autosize from 'autosize';
+import React from 'react';
+
+interface Props extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
+  async: boolean;
+  innerRef?: React.RefObject<HTMLTextAreaElement>;
+  maxRows?: number;
+}
+
+export default class TextareaAutosize extends React.Component<Props> {
+  static readonly defaultProps = {
+    async: false,
+    rows: 1,
+  };
+
+  private lineHeight?: number;
+  private ref = this.props.innerRef ?? React.createRef<HTMLTextAreaElement>();
```

Comment:
should this be a getter instead so the value doesn't potentially get outdated when the props change? (although probably practically impossible? also there are a bunch of other ref-related things that's being set once in componentDidMount and assumed being the same in componentDidUpdate)

Reply:
Refs aren't managed by React other than committing a value to it after render, so messing with a ref wouldn't normally be triggering renders anyway